### PR TITLE
feat(common): add IANA time zone support for date pipe

### LINKED ISSUE #48279: Support IANA time zone for date formatting
### Which @angular/* package(s) are relevant/related to the feature request?

_No response_

### Description

This was once requested #40865 but is now locked due to inactivity. I want to raise this again because IMO supporting `IANA` timezone names natively is a huge improvement, especially now when you can set the default timezone (which is actually an UTC offset, not a timezone) through [`DATE_PIPE_DEFAULT_OPTIONS`](https://angular.io/api/common/DATE_PIPE_DEFAULT_OPTIONS) token.

Why supporting `IANA` is such a big deal? Because of [Daylight Saving Time (DST)](https://www.timeanddate.com/time/dst/statistics.html). 
Countries that are using DST does not have a fixed UTC offset, it depends on the season.
So setting a default timezone (which is an UTC offset) through `DATE_PIPE_DEFAULT_OPTIONS` does not bring that much value today, it's even hard to think of a use case for this. For applications that let the user to set their preferred timezone in settings this token can't be utilized, you're still forced to come up with workaroundish solutions like creating a custom date pipe.

This is a default behaviour when timezone is not provided:

```html
<!-- 
Browser timezone: Europe/Vilnius
DATE_PIPE_DEFAULT_OPTIONS: not provided
-->

{{ '2022-10-29T05:00:00.000Z' | date: 'full' }} // Thursday, 29 October 2022 at 08:00:00 GMT+03:00 ✅ 
{{ '2022-10-31T06:00:00.000Z' | date: 'full' }} // Thursday, 31 October 2022 at 08:00:00 GMT+02:00 ✅ 
<!-- As you can see DST is handled flawlessly. -->
```

Now imagine we want to format all dates the same for all the users not matter on what timezone they are (maybe it's a company timezone, whatever). What `DATE_PIPE_DEFAULT_OPTIONS` value has to be provided for this to work? `UTC+2` or `UTC+3` (in case of `Europe/Vilnius` tz)? You see, you can't choose one, both are needed depending on the date you're trying to format.

```html
<!-- 
Browser timezone: Europe/Stockholm (does not matter, can be any tz)
DATE_PIPE_DEFAULT_OPTIONS: { timezone: 'UTC+2' }
-->

{{ '2022-10-29T05:00:00.000Z' | date: 'full' }} // Thursday, 29 October 2022 at 07:00:00 GMT+02:00 ❌ 
{{ '2022-10-31T06:00:00.000Z' | date: 'full' }} // Thursday, 31 October 2022 at 08:00:00 GMT+02:00 ✅ 
```

```html
<!-- 
Browser timezone: Europe/Stockholm (does not matter, can be any tz)
DATE_PIPE_DEFAULT_OPTIONS: { timezone: 'UTC+3' }
-->

{{ '2022-10-29T05:00:00.000Z' | date: 'full' }} // Thursday, 29 October 2022 at 08:00:00 GMT+03:00 ✅ 
{{ '2022-10-31T06:00:00.000Z' | date: 'full' }} // Thursday, 31 October 2022 at 09:00:00 GMT+03:00 ❌ 
```

### Proposed solution

Add support for [`IANA`](https://www.iana.org/time-zones) time zone name in all these APIs:

https://angular.io/api/common/DATE_PIPE_DEFAULT_OPTIONS
```
{ provide: DATE_PIPE_DEFAULT_OPTIONS, useValue: { timezone: 'Europe/Vilnius' } }
```

https://angular.io/api/common/formatDate
```
{{ myDate | date: 'full':'Europe/Vilnius' }}
```

https://angular.io/api/common/DatePipe
```ts
formatDate(myDate, 'full', 'Europe/Vilnius');
```

### Alternatives considered

If for some reason `IANA` time zone support is **not** considered to be implemented now or in the future then I would suggest renaming the `timezone` property to `utcOffset` which would be more accurate and less confusing for developers.

### PR DESCRIPTION
**Pull request by dzonatan**
## PR Checklist

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

- [x] Feature

## What is the new behavior?

Adds an IANA time zone name (e.g., "Europe/Vilnius") support for the date pipe, which is considered to be more reliable than the UTC offset, mainly because of daylight saving time (DST) respect. 

Closes #48279 

In that issue Jessica also mentions about possibility to fully switch to `Intl` API, which to me looks way out of scope with lots of challenges just for the sake of adding IANA tz support.

The current solution might seem a bit hacky - read the comments in the code on why.
Because of that, I did some manual testing to be sure that `fr` locale uses `UTC` offset (and not some kind of abbreviation like `EET`) for all the time zones in all the platforms (latest Chrome, latest Firefox, latest Safari, and Node.js `14.19.3`) by basically running this and inspecting every value for any weirdness:
```ts
Intl.supportedValuesOf('timeZone')
  .map((timeZone) => 
     Intl.DateTimeFormat('fr', { timeZone, timeZoneName: 'short' })
       .formatToParts(new Date(Date.UTC(2022, 9, 31, 0, 0, 0))).pop()
  );
```

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/common/src/i18n/format_date.ts [hunk 1]
```diff
@@ -718,7 +719,23 @@ function getDateFormatter(format: string): DateFormatter|null {
...
+    // If timezone includes a forward-dash, it's an IANA timezone name (e.g., Europe/Vilnius).
+    // Ideally, we could find out an UTC offset by using the `timeZoneName: 'shortOffset'` option.
+    // Unfortunately, this option is only available from Node.js v17.0.0,
+    // which means it can't be used as long as Angular needs to support older Node.js versions.
+    // Till then, we're left with workaround solutions like such: use a locale (like French),
+    // which uses a short offset that start with "UTC" instead of abbreviation (e.g., EET, CAT, etc)
+    // in all timezones.
+    const shortOffset = Intl.DateTimeFormat('fr', {timeZone: timezone, timeZoneName: 'short'})
+                            .formatToParts(date)
+                            .find(({type}) => type === 'timeZoneName');
```
Comment (dzonatan) [inline packages/common/src/i18n/format_date.ts:733]: One thing that I'm not so sure about is what we should do with invalid values for `timeZone`.
Throw an error? Fallback to the system timezone with a warning in the console?

Currently, it throws an error `Uncaught RangeError: Invalid time zone specified: Non/ExistentName`.

File: packages/common/src/i18n/format_date.ts [hunk 2]
```diff
@@ -718,7 +719,23 @@ function getDateFormatter(format: string): DateFormatter|null {
...
+function timezoneToOffset(date: Date, timezone: string, fallback: number): number {
+  if (timezone.includes('/')) {
+    // If timezone includes a forward-dash, it's an IANA timezone name (e.g., Europe/Vilnius).
+    // Ideally, we could find out an UTC offset by using the `timeZoneName: 'shortOffset'` option.
+    // Unfortunately, this option is only available from Node.js v17.0.0,
+    // which means it can't be used as long as Angular needs to support older Node.js versions.
+    // Till then, we're left with workaround solutions like such: use a locale (like French),
+    // which uses a short offset that start with "UTC" instead of abbreviation (e.g., EET, CAT, etc)
+    // in all timezones.
+    const shortOffset = Intl.DateTimeFormat('fr', {timeZone: timezone, timeZoneName: 'short'})
```
Comment (dylhunn) [inline packages/common/src/i18n/format_date.ts:731]: Why always using fr style dates?

File: packages/common/src/i18n/format_date.ts (hunk 1 above)
Comment (dylhunn) [reply packages/common/src/i18n/format_date.ts:733]: Throwing seems fine to me.

File: packages/common/src/i18n/format_date.ts (hunk 2 above)
Comment (dzonatan) [reply packages/common/src/i18n/format_date.ts:731]: As mentioned in the comments above, it is used because `fr` is one of the locales (which I picked randomly) that uses an offset (`UTC±X` format) with `timeZoneName: 'short'` parameter in all the time zones.
Other locales, like `en-GB`, use abbreviations (e.g., `EET` for `Europe/Vilnius`, `CET` for `Europe/Oslo`, etc.) which is not what we need:
```js
Intl.DateTimeFormat('en-US', { timeZone: `America/Cayman`, timeZoneName: 'short' }).formatToParts(new Date()).pop();
// result: EST
Intl.DateTimeFormat('en-GB', { timeZone: `Europe/Oslo`, timeZoneName: 'short' }).formatToParts(new Date()).pop()
// result: CET

Intl.DateTimeFormat('fr', { timeZone: `America/Cayman`, timeZoneName: 'short' }).formatToParts(new Date()).pop();
// result: UTC-5
Intl.DateTimeFormat('fr', { timeZone: `Europe/Oslo`, timeZoneName: 'short' }).formatToParts(new Date()).pop();
// result: UTC+1
```

Ideally, we should use `timeZoneName: 'shortOffset'` (so all the locales, including `en-US`, would return a proper `UTC/GMT±X` offset and not some kind of abbreviation) but we can't do this today because in Node.js this parameter is only supported from `v17.0.0`. While `timeZoneName: 'short'` is supported from `v0.12.0`.

Nonetheless, there shouldn't be anything wrong in using a fixed locale (in this case `fr`) because all the offsets should be the same in all the locales.

Basically, `timeZoneName: 'short'` acts indentical as `timeZoneName: 'shortOffset'` in `fr` locale:
```js
Intl.DateTimeFormat('fr', { timeZone: 'Europe/Oslo', timeZoneName: 'short' }).formatToParts(new Date()).pop();
// result: UTC+1

 Intl.DateTimeFormat('fr', { timeZone: 'Europe/Oslo', timeZoneName: 'shortOffset' }).formatToParts(new Date()).pop();
// result: UTC+1
```

**Comment by Predhin** [comment]
Do we have a timeline on when this will be merged ?
Thanks.

**Comment by JeanMeche** [comment]
Hi, thank you for proposing a change. As part of #54470 we're currently investigating using the `Intl` globally for I18n. 
Feel free to subscribe to the issue for any updates on that topic.
